test/dotnet-restore.Tests/GivenThatIWantToRestoreApp.cs

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.Build.Construction;
+using Microsoft.DotNet.Tools.Test.Utilities;
+using Xunit;
+using FluentAssertions;
+using System.Linq;
+
+namespace Microsoft.DotNet.Restore.Tests
+{
+    public class GivenThatIWantToRestoreApp : TestBase
+    {
+        [Theory]
+        [InlineData("test1", "test1/", false)]
+        [InlineData("test2", "test2", false)]
+        [InlineData("test3", "test3/abcde", false)]
+        [InlineData("test4", "test4/abcde/", false)]
+        [InlineData("test5", "test5/", true)]
+        [InlineData("test6", "test6", true)]
+        [InlineData("test7", "test7/abcde", true)]
+        [InlineData("test8", "test8/abcde/", true)]
+        public void ItRestorestoSpecificDirectory(string identifier, string dir, bool useFullPath)
+        {
+            var rootPath = TestAssetsManager.CreateTestDirectory(identifier: identifier).Path;
+
+            dir = dir.Replace('/', Path.DirectorySeparatorChar);
+            string fullPath = Path.GetFullPath(Path.Combine(rootPath, dir));
+
+            new NewCommand()
+                .WithWorkingDirectory(rootPath)
+                .Execute()
+                .Should()
+                .Pass();
+
+            string args = $"--packages \"{EscapePath(useFullPath ? fullPath : dir)}\"";
+            new RestoreCommand()
+                 .WithWorkingDirectory(rootPath)
+                 .Execute(args)
+                 .Should()
+                 .Pass()
+                 .And.NotHaveStdErr();
+
+            Directory.Exists(fullPath).Should().BeTrue();
+            Directory.EnumerateFiles(fullPath, "*.dll", SearchOption.AllDirectories).Count().Should().BeGreaterThan(0);
```

Comment:
The intent of the associated issue was broader than what we're testing. Once restore occurs with `--packages`, subsequent calls to CLI should work without additional work. E.g. we should be able to `run`, `build`, `test`, etc. without passing around additional state.

I would reduce this test to a single case [maybe two, as we may not be able to run and test the same project] since the arguments to `--packages` are a nuget concern. Instead, I'd focus the test on our ability to execute the subsequent CLI commands atop this local restore without setting e.g. `NUGET_PACKAGES`. 

This is actually a bit hard to prove deterministically inside a test, since it's difficult to control for a package getting restored into the global nuget cache during the execution of this test. Here's one recipe that seems like it should work:

During the execution of the test:
 - Create a new directory with a Guid.NewGuid() name
 - Create a new directory called `app`
 - Create a new directory called `nupkgs`
 - Create a new directory called `packages`
 - in the guid directory `dotnet new -t lib`
 - in the guid directory `dotnet pack --output {path to `nupkgs`}
 - in `app` do `dotnet new`
 - in `app.csproj` add a packagereference to {GUID}
 - in `app` add a NuGet.config that does not clear out previous configuration but adds `nupkgs` as a source
 - in `app` make sure that Program.cs uses a type from {Guid}
 - restore `app`
 - validate that all the other scenarios work

The GUID-named dependency ensures that the global nuget packages directory does not already contain this package. This avoids wholesale race conditions, artifact re-use etc. You can probably simplify the consuming app a bit via a well-crafted Test Asset {e.g. one that takes /p:GuidPackageName so you don't need to modify it at restore time}. Using a type from the dependency package ensures that the library delivered by the NuGet package can actually be found.